### ISSUE TITLE
Constructor injection for Blazor components

### ISSUE THREAD

**Issue by SQL-MisterMagoo** [dotnet/aspnetcore#18088]
### Is your feature request related to a problem? Please describe.
Constructor injection in Components was ruled out previously as "we don't plan to do this" but now that we have partials is there any change in this attitude?
see #15779 , #5497

### Describe the solution you'd like
If Blazor could be modified to optionally use DI to source components, then constructor injection would be possible,

If you would be open to at least reviewing a PR, I'm sure the community would do the groundwork to make it happen.

**Comment by pranavkm** [comment]
Thanks for your issue report. We'll track fixing this as part of our next milestone.

**Comment by SQL-MisterMagoo** [comment]
Thanks for that

**Comment by nvmkpk** [comment]
This is a must have. Property based injection prevents us from consuming them within the constructor there by preventing us from making some (other) fields read only.

**Comment by YairHalberstadt** [comment]
Comment from a duplicate issue I opened:

Currently blazor only allows injecting services into components using property injection via the `[Inject]` attribute.

However this has a number of disadvantages over constructor injection

1. NRTs
The compiler warns if you declare you service as non-nullable, and nullable reference types are enabled. You have to manually suppress the warning.
2. Immutablility
You cannot declare the services readonly, since they need to be set by the injector.
3. Manual construction
If you create the component manually (perhaps for unit testing), there is nothing to guarentee you will set the services. Even worse - you often can't since the properties are usually private. You have to declare two constructors - one parameterless, and one which sets all the services. This is needless code duplication, and means that test and app code go through different paths, which is undesirable.
4. Validation
The constructor is a place where you can validate all your services, and do transformations on them. Perhaps, you don't want to store your services, but just call one service with another service, and store the result. Constructors are a good place to do that, with property injection this is tricky.

### Proposal
If there is a single constructor, use it for dependency injection. It is an error if any of the parameters cannot be resolved. It is an error if there are multiple constructors, none of which are parameterless, unless one of them is marked with `[Inject]`. After the construction, property injection takes place.

**Comment by nvmkpk** [comment]
`InjectAttribute` cannot be set on non-properties. There is already `ActivatorUtilitiesConstructorAttribute` available for decorating a constructor to be used by DI.

**Comment by janissimsons** [comment]
Any updates on this?

**Comment by mkArtakMSFT** [comment]
Moving to backlog given this now be achieved using `IComponentActivator`.

**Comment by janissimsons** [comment]
@mkArtakMSFT how can it be achieved using IComponentActivator? Could you please give an example?

**Comment by hikalkan** [comment]
@janissimsons,
 
As I can see, `IComponentActivator` is a service that is used when the renderer needs to create a new component instance.
The default implementation is [DefaultComponentActivator](https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/DefaultComponentActivator.cs) which uses the `Activator.CreateInstance`.

So, you can just implement the `IComponentActivator`, inject the `IServiceProvider` into your implementation and use the `.GetService(Type)` method to create the compoent. I suggest to fallback the `Activator.CreateInstance` if GetService returns null (that means the component hasn't registered to the DI).

Then you need to register your components to the DI.

**Comment by hikalkan** [comment]
The implementation can be like that:

````csharp
using System;
using Microsoft.AspNetCore.Components;

namespace MyProject
{
    public class ServiceProviderComponentActivator : IComponentActivator
    {
        public IServiceProvider ServiceProvider { get; }

        public ServiceProviderComponentActivator(IServiceProvider serviceProvider)
        {
            ServiceProvider = serviceProvider;
        }

        public IComponent CreateInstance(Type componentType)
        {
            var instance = ServiceProvider.GetService(componentType);

            if (instance == null)
            {
                instance = Activator.CreateInstance(componentType);
            }

            if (!(instance is IComponent component))
            {
                throw new ArgumentException($"The type {componentType.FullName} does not implement {nameof(IComponent)}.", nameof(componentType));
            }

            return component;
        }
    }
}
````

Then replace the `IComponentActivator` service: 

````csharp
services.Replace(ServiceDescriptor.Transient<IComponentActivator, ServiceProviderComponentActivator>());
````

@mkArtakMSFT is there any problem with this implementation? Does it cause a memory leak (I don't know how the IServiceProvider scope is managed for this case)?

**Comment by nvmkpk** [comment]
Keep in mind that this only works in .net 5. I wish this was the done by default. But this will do it for me.

**Comment by TanvirArjel** [comment]
Constructor DI should be default in Blazor and it can be a very good feature for **6.0.0**

**Comment by TanvirArjel** [comment]
I have created a library called `TanvirArjel.Blazor` which enable constructor dependency injection support for Blazor Components with a single line of code as follows:

*Blazor Server:*

```C#
using TanvirArjel.Blazor.DependencyInjection;
    
services.AddComponents();
```
*Blazor Web Assembly:*

```C#
using TanvirArjel.Blazor.DependencyInjection;
    
builder.Services.AddComponents();
```

Install the `TanvirArjel.Blazor` [nuget](https://www.nuget.org/packages/TanvirArjel.Blazor/) package into your Blazor app  as follows:

**PMC:**

     Install-Package TanvirArjel.Blazor
     
**.NET CLI:**

     dotnet add package TanvirArjel.Blazor

**Comment by TLabWest** [comment]
Because Nullable, our component classes are riddled with unsightly workarounds like:

```
SomeType? _theProp;
[Parameter] SomeType TheProp
{
  get => _theProp ?? throw new Exception($"{nameof(TheProp)} is null");
  set => _theProp = value;
}
```

Constructor injection would clean up so much mess.

**Comment by andrewrondeau-optirtc** [comment]
FYI: Swift lets you do something like:

`[Parameter] public SomeType! TheProp`

It's functionally equivalent to @TLabWest 's example, just less typing.

**Comment by pranavkm** [comment]
Support for constructor injection was merged as part of .NET 7 preview2. That said it doesn't help with parameter binding that some of the later comments discuss. I'd recommend filing a separate issue to discuss it.

**Comment by kaleidocore** [comment]
> FYI: Swift lets you do something like:
> 
> `[Parameter] public SomeType! TheProp`
> 
> It's functionally equivalent to @TLabWest 's example, just less typing.

I admittedly don't know exactly what code Swift would generate behind the scenes but I suspect it's not quite equivalent. You _could_ simply use the null forgiving operator `!` like:

`[Parameter] public SomeType TheProp = null!`

But this would eventually lead to an obscure NullRef exception, rather than a message with explicit information to set the required parameter.

I know there has been some parallel discussion on a "required" option for component parameters. Perhaps that could somehow be tied to the constructor parameters?

**Comment by TanvirArjel** [comment]
Why this has been reopened? Anything went wrong?

**Comment by pranavkm** [comment]
Unfortunately we had to roll back this change because the change in the current format results in size regression in Blazor WebAssembly. We'll revisit this change once additional runtime changes are available that do not regress the size. For more context, see https://github.com/dotnet/aspnetcore/issues/40521. 

In the meanwhile, the plan is for us to document using `IComponentActivator` to configure within the app.

**Comment by danroth27** [comment]
We think this is currently blocked on https://github.com/dotnet/runtime/issues/66153

**Comment by steveharter** [comment]
@danroth27 @mkArtakMSFT 

> We think this is currently blocked on https://github.com/dotnet/runtime/issues/66153

That is now unblocked. There is still perf research\work to do here on the DI side, but the size regression is no longer blocked. There is now a reflection path that doesn't use LINQ.

**Comment by Dargazo** [comment]
> Comment from a duplicate issue I opened:
> 
> Currently blazor only allows injecting services into components using property injection via the `[Inject]` attribute.
> 
> However this has a number of disadvantages over constructor injection
> 
>     1. NRTs
>        The compiler warns if you declare you service as non-nullable, and nullable reference types are enabled. You have to manually suppress the warning.
> 
>     2. Immutablility
>        You cannot declare the services readonly, since they need to be set by the injector.
> 
>     3. Manual construction
>        If you create the component manually (perhaps for unit testing), there is nothing to guarentee you will set the services. Even worse - you often can't since the properties are usually private. You have to declare two constructors - one parameterless, and one which sets all the services. This is needless code duplication, and means that test and app code go through different paths, which is undesirable.
> 
>     4. Validation
>        The constructor is a place where you can validate all your services, and do transformations on them. Perhaps, you don't want to store your services, but just call one service with another service, and store the result. Constructors are a good place to do that, with property injection this is tricky.
> 
> ...
Looks like this is going to make it into .NET 9, which is great!

In the meantime, items 1, and 3 can be addressed by marking the service properties 'public required' and e.g. use an 'init' setter. Although not ideal, it might be preferable to suppressing the nullability warnings

**Comment by andrewrondeau-optirtc** [comment]
> In the meantime, items 1, and 3 can be addressed by marking the service properties 'public required' and e.g. use an 'init' setter. Although not ideal, it might be preferable to suppressing the nullability warnings

Yes, that is the best workaround. I was setting the properties to null! (notice the exclamation mark) until I discovered the "required" keyword.
